### PR TITLE
Improvements on the license model implementation

### PR DESCRIPTION
**Pull request by blaumeiser-at-bosch**
#### Major changes:
- Add method to return in LicenseStatement the next level of the expression, this enables the traversal through a license expression
- Reimplement the semantics of a with license with details:
  - Change the semantics to represent one license as WithLicense with two parts
  - Base WithLicense on a single license, all methods of License are adapted to meet the needs of a with license
  - The basic concept of a with license is that two SPDX identifier are matched by the term ' WITH ', this is also true for the name of the license not only the id
  - The license text is also attached to each other in the form of license text with exception: exception text and corresponding newlines
  - WithLicense can be built from two parts reflecting original license and exception and from already combined terms, e.g., 'licenseID WITH exceptionID
    - Corresponding Constructors and setters have been defined, getters are added for both as convenience 
  - Tests are added to check for the new behavior
  - Adapt places in the code that requires a change due to the changed understanding of WithLicense

#### Minor changes:
- Replace use of List by use of Collection, because a LicenseStatement is a commutative operation, so order must not be important
- Improve parameter handling to ensure non-null parameters, this prevents NullPointerExceptions in use due to immediate feedback
- Small improvement of local stream handling, using the joining collector with pre- and suffix
- Adaptations of code places where interface changes mattered

### Request Reviewer
@neubs-bsi @sschuberth 

### Type of Change
*Type of change*:  improvement

### How Has This Been Tested?
Unittests have been added, tests have been adapted to reflect the changed semantics

### Checklist
Must:
- [x] All related issues are referenced in commit messages

Optional: *(delete if not applicable)*
- [x] I have provided tests for the changes (if there are changes that need additional tests)
- [x] I have updated the documentation accordingly to my changes as javadoc

### CODE REVIEW

File: core/core-workflow-steps/src/main/java/org/eclipse/sw360/antenna/workflow/processors/LicenseKnowledgeBaseResolverImpl.java [hunk 1]
```diff
@@ -46,7 +46,7 @@ public LicenseKnowledgeBaseResolverImpl(ILicenseManagementKnowledgeBase knowledg
                 .flatMap(List::stream)
                 .map(ArtifactLicenseInformation::get)
                 .map(LicenseInformation::getLicenses)
-                .flatMap(List::stream)
+                .flatMap(Collection::stream)
```
Comment (sschuberth) [inline core/core-workflow-steps/src/main/java/org/eclipse/sw360/antenna/workflow/processors/LicenseKnowledgeBaseResolverImpl.java]: Just a general comment: Using bullet points in a commit message is a good hint that you're probably doing more in a single commit than you should do. If you need to enumerate things, move each point into a separate commit instead.

File: core/model/src/test/java/org/eclipse/sw360/antenna/model/license/LicenseTest.java [hunk 2]
```diff
@@ -39,6 +39,15 @@ public void testLicenseWithOnlyId() {
...
 
+    @Test
+    public void testEqualLicense() {
+        License license1 = new License ("Apache-2.0");
+        License license2 = new License ("Apache-2.0", "Apache License Version 2.0", "Some license text");
+        License license3 = new License ("MIT");
+
+        assertThat(license1.equalLicense(license2)).isTrue();
+        assertThat(license1.equalLicense(license3)).isFalse();
+    }
```
Comment (sschuberth) [inline core/model/src/test/java/org/eclipse/sw360/antenna/model/license/LicenseTest.java]: Nit: Missing blank line after this one.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/FromXmlLicenseInformationBuilder.java [hunk 3]
```diff
@@ -114,7 +114,7 @@ public void setException(LicenseBuilder exception) {
 
         @Override
         public LicenseInformation build() {
-            return new WithLicense((License) license.build(), (License) exception.build());
+            return new WithLicense(license.id, license.commonName, license.text, exception.id, exception.commonName, exception.text);
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/license/FromXmlLicenseInformationBuilder.java]: A general remark: We should watch out to not re-implement too much of either https://github.com/oss-review-toolkit/ort/tree/master/spdx-utils/src/main/kotlin or https://github.com/spdx/Spdx-Java-Library/tree/master/src/main/java/org/spdx/library/model/license here.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java [hunk 4]
```diff
@@ -23,83 +34,167 @@
...
 
+    /**
+     * Return the value of a property or an empty Optional.
+     *
+     * @param key The key of the queried property, must not be null.
+     * @return The property value stored or an empty Optional.
+     */
     public Optional<String> getProperty(String key) {
-        return Optional.ofNullable(properties.get(key));
+        return Optional.ofNullable(properties.get(requireNonNull(key)))
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java]: This is more than just adding comments.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java [hunk 5]
```diff
@@ -130,8 +225,15 @@ public boolean equals(Object o) {
...
+    /**
+     * Compares the license to another license. The licenses are equals, if the license ids are the same,
+     * independent whether one license is complete or not.
+     *
+     * @param license The license to compare this license to.
+     * @return True, if both licenses have the same license id.
+     */
     public boolean equalLicense(License license) {
-        return Objects.equals(evaluate(), license.evaluate());
+        return Objects.equals(getId(), license.getId());
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java]: This is more than just adding comments.

File: core/model/src/test/java/org/eclipse/sw360/antenna/model/license/WithLicenseTest.java [hunk 6]
```diff
@@ -88,6 +88,33 @@ public void testWithLicenseBuildFromAssembledInput() {
         assertThat(withLicense1.equalLicense(withLicense2)).isTrue();
     }
 
+    @Test(expected=IllegalArgumentException.class)
```
Comment (sschuberth) [inline core/model/src/test/java/org/eclipse/sw360/antenna/model/license/WithLicenseTest.java]: A general remark: According to the commit message, only this file should be changes, and all changes in other files should be separated out to other commit with proper commit messages.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java (hunk 4 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java]: I do not get your point, the commit message also says, fix code smells, the fixes below are doing that, fixing the code smells that were reported by Sonarqube.

File: core/model/src/test/java/org/eclipse/sw360/antenna/model/license/WithLicenseTest.java (hunk 6 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/test/java/org/eclipse/sw360/antenna/model/license/WithLicenseTest.java]: See my answer above, the other stuff were code smells reported by SonarQube which I fixed as the commit message says. You could argue, that both things could be separated into two commits, but I do not see here the advantage of having this separated, because the changes are not really changing a lot.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java (hunk 5 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/license/License.java]: I'll change this.

File: core/model/src/test/java/org/eclipse/sw360/antenna/model/license/WithLicenseTest.java (hunk 6 above)
Comment (sschuberth) [reply core/model/src/test/java/org/eclipse/sw360/antenna/model/license/WithLicenseTest.java]: The advantage of separating semantically different changes into different commits for me as a reviewer is that I can skip over commits that say to change things which are either unimportant to me, or that are things I cannot comment on.

So, if a commit says to just add a test, that hardly breaks the main app, and I might skip the review if I'm short of time. Similarly, if a commit says to just add comments, that is a non-functional change which I can just skim over and proceed to the next commit.

However, for that to work I need to be able to rely on the commit message to really only change what is stated here. At least when the commit comes from a trusted contributor. For untrusted parties I of course review everything, thoroughly :-)

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/FromXmlLicenseInformationBuilder.java (hunk 3 above)
Comment (blaumeiser-at-bosch) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/license/FromXmlLicenseInformationBuilder.java]: In principle I agree, but there has to be a step-by-step move towards the intended solution. In the meanwhile, we need this step to continue with "smaller" steps.

**Comment by blaumeiser-at-bosch** [comment]
@sschuberth This is it, no commit found, that could be smaller.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/license/LicenseStatement.java [hunk 7]
```diff
@@ -10,10 +10,7 @@
  */
 package org.eclipse.sw360.antenna.model.license;
 
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.List;
-import java.util.Objects;
+import java.util.*;
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/license/LicenseStatement.java:13]: Nit: The commit message just repeats in plain word what's obvious from the diff, but lacks context *why* this change makes sense.

File: core/runtime/src/test/java/org/eclipse/sw360/antenna/configuration/ConfigurationTest.java [hunk 8]
```diff
@@ -62,7 +62,7 @@ public void test() {
 
         ArtifactIdentifier identifier = new ArtifactFilename("overrideName.jar");
         assertThat(configuration.getFinalLicenses().keySet().stream().anyMatch(k -> k.matches(identifier))).isTrue();
-        List<LicenseInformation> list = configuration.getFinalLicenses().entrySet().stream()
+        List<License> list = configuration.getFinalLicenses().entrySet().stream()
```
Comment (sschuberth) [inline core/runtime/src/test/java/org/eclipse/sw360/antenna/configuration/ConfigurationTest.java:65]: To the reader who's not so familiar with the code base (like me) it's surprising that this actually just works. It would be nice to read in the commit message *why* this is a fix, and why it "worked" (i.e. compiled) also before, but what's better now.

**Comment by sschuberth** [comment]
> @sschuberth This is it, no commit found, that could be smaller.

Thank you, this is *much* more pleasant to review now!
